cpp/src/lists/apply_boolean_mask.cu

```diff
@@ -0,0 +1,140 @@
+
+class get_list_size {
+ public:
+  explicit get_list_size(lists_column_view const& lcv)
+    : num_rows{lcv.size()},
+      offsets{lcv.offsets().begin<offset_type>() + lcv.offset()},
+      bitmask{lcv.null_mask()}
+  {
+  }
+
+  size_type __device__ operator()(size_type i) const
+  {
+    return bit_value_or(bitmask, i, true) ? (offsets[i + 1] - offsets[i]) : 0;
+  }
+
+ private:
+  size_type num_rows;
+  offset_type const* offsets;
+  bitmask_type const* bitmask;
+};
+
+void assert_same_list_sizes(lists_column_view const& input,
+                            lists_column_view const& boolean_mask,
+                            rmm::cuda_stream_view stream)
+{
+  auto const begin = cudf::detail::make_counting_transform_iterator(
+    0,
+    [get_list_size = get_list_size{input}, get_mask_size = get_list_size{boolean_mask}] __device__(
+      size_type i) -> size_type { return get_list_size(i) != get_mask_size(i); });
+
+  CUDF_EXPECTS(thrust::reduce(rmm::exec_policy(stream), begin, begin + input.size()) == 0,
+               "Each list row must match the corresponding boolean mask row in size.");
+}
+}  // namespace
+
+std::unique_ptr<column> apply_boolean_mask(lists_column_view const& input,
+                                           lists_column_view const& boolean_mask,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(input.size() == boolean_mask.size(),
+               "Boolean masks column must have same number of rows as input.");
+
+  auto const num_rows = input.size();
+
+  if (num_rows == 0) { return cudf::empty_like(input.parent()); }
+  // Note: This assert guarantees that no elements are gathered
+  // from nominally NULL input list rows.
+  assert_same_list_sizes(input, boolean_mask, stream);
+
+  auto constexpr offset_data_type = data_type{type_id::INT32};
+
+  auto filtered_child = [&] {
+    std::unique_ptr<cudf::table> tbl =
+      cudf::detail::apply_boolean_mask(cudf::table_view{{input.get_sliced_child(stream)}},
+                                       boolean_mask.get_sliced_child(stream),
+                                       stream,
+                                       mr);
+    std::vector<std::unique_ptr<cudf::column>> columns = tbl->release();
+    return std::move(columns.front());
+  };
+
+  auto output_offsets = [&] {
+    auto boolean_mask_sliced_offsets =
+      cudf::detail::slice(
+        boolean_mask.offsets(), {boolean_mask.offset(), boolean_mask.size() + 1}, stream)
+        .front();
+
+    auto const sizes         = cudf::reduction::segmented_sum(boolean_mask.get_sliced_child(stream),
+                                                      boolean_mask_sliced_offsets,
+                                                      offset_data_type,
+                                                      null_policy::EXCLUDE,
+                                                      stream);
```

Comment:
Is this the same as the way we typically compute offsets with `thrust::exclusive_scan`?

Reply:
The aim here is to figure out the sizes of each output list row, based on the number of elements selected by the boolean mask:
E.g. Consider this input `list<bool>`:
```c++
[ {0,1,0,1}, {1,1}, {1,X,1}, {1}, {} ]
```
This should yield the following output sizes:
```c++
[ 2, 2, 2, 1, 0 ]
```

I'm unable to see how `thrust::exclusive_scan()` might achieve what we need here:
1. It has no notion of nulls. We would need to use an optional iterator or something.
2. It has no notion of segments.

The `cudf::reduction::segmented_*()` algorithms already address this more directly, and with corner-case handling.

Reply:
Technically you could do it with a scan_by_key and a null_replacement iterator, but I don't have an issue with this approach.

Reply:
Thanks for reminding me of `null_replacement_iterator`. This helped avoid manifesting an intermediate result column.